qa/util/src/main/java/io/camunda/zeebe/qa/util/junit/ZeebeIntegrationExtension.java

```diff
@@ -0,0 +1,345 @@
+      return directory;
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private ClusterResource asClusterResource(final Object testInstance, final Field field) {
+    final TestCluster value;
+
+    try {
+      value = (TestCluster) ReflectionUtils.makeAccessible(field).get(testInstance);
+    } catch (final IllegalAccessException e) {
+      throw new UnsupportedOperationException(e);
+    }
+
+    return new ClusterResource(value, field.getAnnotation(TestZeebe.class));
+  }
+
+  private ApplicationResource asNodeResource(final Object testInstance, final Field field) {
+    final TestApplication<?> value;
+
+    try {
+      value = (TestApplication<?>) ReflectionUtils.makeAccessible(field).get(testInstance);
+    } catch (final IllegalAccessException e) {
+      throw new UnsupportedOperationException(e);
+    }
+
+    return new ApplicationResource(value, field.getAnnotation(TestZeebe.class));
+  }
+
+  private Store store(final ExtensionContext extensionContext) {
+    return extensionContext.getStore(Namespace.create(ZeebeIntegrationExtension.class));
+  }
+
+  private record ClusterResource(TestCluster cluster, TestZeebe annotation)
+      implements TestZeebeResource, CloseableResource {
+
+    @Override
+    public void close() {
+      CloseHelper.close(
+          error -> LOG.warn("Failed to close cluster {}, leaking resources", cluster.name(), error),
+          cluster);
+    }
+
+    @Override
+    public void start() {
+      cluster.start();
+    }
+
+    @Override
+    public void await(final TestHealthProbe probe) {
+      cluster.await(probe);
+    }
+
+    @Override
+    public void awaitCompleteTopology() {
+      final var clusterSize =
+          annotation.clusterSize() <= 0 ? cluster.brokers().size() : annotation.clusterSize();
+      final var partitionCount =
+          annotation.partitionCount() <= 0
+              ? cluster.partitionsCount()
+              : annotation.partitionCount();
+      final var replicationFactor =
+          annotation.replicationFactor() <= 0
+              ? cluster.replicationFactor()
+              : annotation.replicationFactor();
+      cluster.awaitCompleteTopology(
+          clusterSize, partitionCount, replicationFactor, Duration.ofMinutes(clusterSize));
+    }
+  }
+
+  private record ApplicationResource(TestApplication<?> app, TestZeebe annotation)
+      implements TestZeebeResource, CloseableResource {
+
+    @Override
+    public void close() {
+      CloseHelper.close(
+          error -> LOG.warn("Failed to close test app {}, leaking resources", app.nodeId()), app);
+    }
+
+    @Override
+    public void start() {
+      app.start();
+    }
+
+    @Override
+    public void await(final TestHealthProbe probe) {
+      app.await(probe);
+    }
+
+    @Override
+    public void awaitCompleteTopology() {
+      if (!(app.isGateway() && (app instanceof final TestGateway<?> gateway))) {
```

Comment:
:question: Shouldn't this work for embedded gateways too? So just baling out of `isGateway` returns false would be enough, no instanceof check needed?

Reply:
Technically yes, but doing the cast check allows me to use the `gateway` variable further down as a `TestGateway`, so I figured it doesn't hurt ;)

Reply:
Oh yeah that makes sense now :+1: 
The missing piece for me was that `TestGateway` is an interface, not the gateway implementation class. All good then, makes sense to me :+1: